### ISSUE TITLE
Typescript bindings

### ISSUE THREAD

**Issue by xavdid** [zapier/zapier-platform#8]
This won't happen until we've got a 1.0 and a stable API, but bindings would be great. I'd be hapy to jump on it once we're ready!

All it'll be a `.ts` file in the repo and a line in the `package.json`

**Comment by philonor** [comment]
@xavdid We are experiencing problems when trying to use Zapier's own error types as described [here](https://platform.zapier.com/cli_docs/docs#halting-execution) in combination with TypeScript.

The described code example `throw new z.errors.HaltedError('Your reason.');` does not work with TypeScript for us. Are we missing something?
- ```
   import {ZObject} from 'zapier-platform-core';
   throw new ZObject.errors.HaltedError('Your reason'). // TS2693: 'ZObject' only refers to a type, but is being used as a value here.
  ```
- ```
   import {HaltedError} from 'zapier-platform-core';
   throw new HaltedError('Your reason'); // passes with tsc, but yields "zapier_platform_core_1.HaltedError is not a constructor" error when run.
  ```
- ```
   import {HaltedError} from 'zapier-platform-core';
   throw new Error ('Your reason.') as HaltedError; // Passes tsc and works when run, but HaltedError is undefined
   ```
Reproduction repository: https://github.com/philonor/zapier-platform/tree/master

https://github.com/philonor/zapier-platform/commit/7fd74fd9d1aa4f9f88ad724af354d9a381e67af6

For now, we can only use default `Error` type which will turn off the Zap after a while instead of halting execution ([source](https://zapier.github.io/zapier-platform/#error-handling)):

> Note that if a Zap raises too many error messages it will be automatically turned off, so only use these if the scenario is truly an error that needs to be fixed.

**Comment by xavdid** [comment]
@philonor the issue is that `ZObject` is just a type, not the actual object. Take the following:

```ts
import {ZObject} from 'zapier-platform-core'
const perform = (z: ZObject, ) => {
  throw z.errors.HaltedError('something wrong')
}
```

Once you type `z.errors.H`, `HaltedError` should auto-complete, thanks to marking `z` as a `ZObject`. Does that make sense?

**Comment by sampsonbryce** [comment]
@xavdid We are having issues with the perform type when testing triggers that use webhooks. The type of createAppTester is
```javascript
export const createAppTester: (
  appRaw: object,
  options?: { customStoreKey?: string }
) => <T extends any, B extends Bundle>(
  func: (z: ZObject, bundle: B) => Promise<T>,
  bundle?: Partial<B> // partial so we don't have to make a full bundle in tests
) => T extends Promise<T> ? T : Promise<T>
```

Which requires that the function passed to `appTester` returns a promise. But for webhooks we don't have to return a promise as described in the documentation here: https://platform.zapier.com/cli_tutorials/resthooks#step-3-write-the-perform-function

So in my perform function I just return the cleaned response as an array 
```javascript
export const perform = (z: ZObject, bundle: Bundle): Array<{ message: string}> => {
  return [bundle.cleanedRequest];
};
```

Now if I import the perform function and pass it to `appTester` I get a type error
```javascript
import App from "../../index";
import { perform } from "../../triggers/activity_entry_created";

const appTester = zapier.createAppTester(App);

describe("Trigger - do something", () => {
  zapier.tools.env.inject();

  describe("perform", () => {
    const bundle = {};

    it("should get an array", async () => {
      // Gives type error: Argument of type '(z: ZObject, bundle: Bundle<{ [x: string]: any; }>) => { message: string; }[]' is not assignable to parameter of type '(z: ZObject, bundle: any) => Promise<{ message: string; }[]>'.
      //                   Type '{ message: string; }[]' is missing the following properties from type 'Promise<{ message: string; }[]>': then, catch, [Symbol.toStringTag], finallyts(2345)
      const results = await appTester<Array<{ message: string }>, any>(perform, bundle);
      results.should.be.an.Array();
    });
  });
});
```

However, this problem is solved by changing the `createAppTester` type to

```javascript
export const createAppTester: (
  appRaw: object,
  options?: { customStoreKey?: string }
) => <T extends any, B extends Bundle>(
  func: (z: ZObject, bundle: B) => Promise<T> | T, // <- CHANGE HERE
  bundle?: Partial<B> // partial so we don't have to make a full bundle in tests
) => T extends Promise<T> ? T : Promise<T>
```

That way the types match up. Otherwise I would have to change my perform function to 
```javascript
export const perform = (z: ZObject, bundle: Bundle): Promise<Array<{ message: string}>> => {
  return Promise.resolve([bundle.cleanedRequest]);
};
```

Which doesn't match the documentation that I linked above at all.

I think the `createAppTester` function argument return type needs to support returning non promise types for webhooks.

**Comment by xavdid** [comment]
@sampsonbryce Thanks for the detailed report!

I think the confusion here is that `appTester` always returns a promise (even if the original method didn't). 

If your test is:

```ts
it("should get an array", async () => {
  const results = await appTester<Promise<Array<{ message: string }>>, any>(perform, bundle);
  results.should.be.an.Array();
});
```

It looks a little silly, but then it more accurately describes what's coming out of the `appTester` call.

**Comment by sampsonbryce** [comment]
@xavdid I don't think that will work and it doesn't seem to work when testing it. The type provided as the first template type to `appTester` is nested in a `Promise` type via the `appTester` type definition. That means that your solution expects a perform function with the type `Promise<Promise<Array<{ message: string }>>>`

I have made a test repo to reproduce the issue so you can see it for yourself. I still think the issue lies in the fact that the `appTester` type requires that the perform function return a `Promise` even though that is not the case

**Comment by xavdid** [comment]
Got it, thank you for that context! i'll take a look next week.

**Comment by xavdid** [comment]
@sampsonbryce ok, thanks for that thorough explanation! I had some time to play around with this today and your change is a sound one! I also cleaned up that definition a little. It'll go out in the next release. 

see #172

**Comment by zomars** [comment]
I can't find anything online so I'll ask here, Is there a way to generate type definitions from Zapier schemas? I've been manually doing it so far but I'm pretty sure there has to be a better way that I'm missing. Any help is appreciated 🙏🏽

**Comment by xavdid** [comment]
@zomars There's not currently a working way, no. I made a lot of progress in https://github.com/zapier/zapier-platform/pull/260, but I was having trouble nailing down the definitions for the functions, which IMO was the most important part. 

For now, referencing the [schema](https://github.com/zapier/zapier-platform/blob/master/packages/schema/docs/build/schema.md) and using the `zapier validate` CLI command is probably the way to go, unfortunately.

**Comment by zomars** [comment]
Thanks @xavdid! I've ended up using the gen script on your branch and manually filling the missing definitions. Definitely saved a lot of time. I could make a PR to the typescript example if you want.

**Comment by xavdid** [comment]
Very cool! Feel free to open a PR or show those to us if you like them. It's been a tough nut to crack

**Comment by Jackman3005** [comment]
I think it would be great if the Zapier Schema package would just convert the [exported-schema.json](https://github.com/zapier/zapier-platform/blob/main/packages/schema/exported-schema.json) to typescript types using [json-schema-to-typescript](https://github.com/bcherny/json-schema-to-typescript#readme) and include it alongside the `exported-schema.json` file that they commit into the repository. This would keep the typescript types completely in sync with the schema and would avoid all end-users of zapier libraries from having to regenerate the typescript types based on the same file all the time (since we won't know if we need to update it or not).
